[many ports] Not supported on staticcrt

Thomas1664 opened the pull request:
**Describe the pull request**

- #### What does your PR fix?
  Many ports are not supported on `staticcrt`. This was expressed in our CI baseline before. I added `"supports": "!staticcrt"` to better express that those ports are not supported when building with static CRT and removed the affected ports from CI baseline.

I opened a single PR for all ports because I think this results in less CI time since some ports depend on LLVM.

Thomas1664 committed:
xalan-c

Thomas1664 committed:
treehopper

Thomas1664 committed:
torch-th

Thomas1664 committed:
qpid-proton

Thomas1664 committed:
pmdk

Thomas1664 committed:
openvr

Thomas1664 committed:
ngspice

Thomas1664 committed:
moos-core

Thomas1664 committed:
monkeys-audio

Thomas1664 committed:
milerius...

Thomas1664 committed:
mesa

Thomas1664 committed:
libplist

Thomas1664 committed:
libmodman

Thomas1664 committed:
libepoxy

Thomas1664 committed:
hpx

Thomas1664 committed:
halide

Thomas1664 committed:
Update CI baseline

Thomas1664 committed:
format

Thomas1664 committed:
version

Thomas1664 committed:
fix treehopper

Thomas1664 committed:
version

Thomas1664 commented:
CC @BillyONeal For the issue with PR suggestion bot at the first CI run

Neumann-A commented on ports/hpx/vcpkg.json:
```diff
@@ -1,13 +1,14 @@
   ],
   "homepage": "https://github.com/STEllAR-GROUP/hpx",
   "license": "BSL-1.0",
+  "supports": "!(windows & staticcrt)",
```
this is incorrect.

Thomas1664 replied:
Why?

dg0yt commented on ports/halide/portfile.cmake:
```diff
@@ -1,11 +1,9 @@
 vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
 
-set(HALIDE_VERSION_TAG v14.0.0)
```
This variable is used in `usage.in`.

dg0yt reviewed (changes requested):
Did you actually check the reasons for failing in `x64-windows-static`, i.e. whether it is really the static CRT, or just poor handling of transitive usage requirments (static library linkage)? 
(I see that there seems to be no record for `x64-windows-static-md` but maybe these ports just cascade in that triplet.)

Thomas1664 committed:
[halide] Restore variable

Thomas1664 committed:
version

Thomas1664 commented:
> Did you actually check the reasons for failing in `x64-windows-static`, i.e. whether it is really the static CRT, or just poor handling of transitive usage requirments (static library linkage)? (I see that there seems to be no record for `x64-windows-static-md` but maybe these ports just cascade in that triplet.)

Yes, they all produce the following output, claiming to not support static CRT linkage:

```
-- Note: ngspice only supports dynamic library linkage. Building dynamic library.
CMake Error at scripts/cmake/vcpkg_check_linkage.cmake:25 (message):
  Refusing to build unexpected dynamic library against the static CRT.

      If this is desired, please configure your triplet to directly request this configuration.
Call Stack (most recent call first):
  ports/ngspice/portfile.cmake:1 (vcpkg_check_linkage)
  scripts/ports.cmake:147 (include)
```

or

```
CMake Error at scripts/cmake/vcpkg_check_linkage.cmake:32 (message):
  mesa only supports dynamic crt linkage
Call Stack (most recent call first):
  ports/mesa/portfile.cmake:12 (vcpkg_check_linkage)
  scripts/ports.cmake:147 (include)
```

Thomas1664 committed:
Fix quote

Thomas1664 committed:
torch-th doesn't provide any config

Thomas1664 committed:
version

Neumann-A commented:
@Thomas1664  reread the warning completly. Your assesment is wrong.

Thomas1664 commented:
> @Thomas1664 reread the warning completly. Your assesment is wrong.

Which warning do you mean? Where can I find the warning?

Again, this is not helpful. Please be more specific!

Neumann-A commented:
> Where can I find the warning?

https://github.com/microsoft/vcpkg/pull/26948#issuecomment-1256582024

especially the line:
`If this is desired, please configure your triplet to directly request this configuration.`

and the wording in 
`Refusing to build unexpected dynamic library against the static CRT.`

This is a vcpkg decision to not build dynamic libraries with static CRT linkage unless the user **explicitly** requests it due to all the gotchas you could have at the dll boundary. But this doesn't mean static CRT linkage is unsupported it is just the same reason why there is no official `x64-windows-mt` triplet.

Thomas1664 commented:
These aren't warnings, these are errors. It literally says `CMake Error`. All of the ports I fixed literally have `vcpkg_check_linkage(ONLY_DYNAMIC_CRT ONLY_DYNAMIC_LIBRARY)`, claiming to only support dynamic CRT linkage. I really don't understand why it should be wrong in this case to express this inside `supports`. If one is OK with the fact that those ports are built with dynamic CRT linkage although static was requested, they could use `--allow-unsupported` and do what is suggested by `If this is desired, please configure your triplet to directly request this configuration.`.

Furthermore, there already was an issue about that: #26661. It was initially fixed by the maintainers in exactly the same way in #26700, except @JackBoosY was following your request of removing `!staticcrt` because there were no errors. However, this is not the case here. All ports I fixed in this PR produce a build error, which is the reason why `supports` should be preferred over CI baseline since **CI baseline should only be used for things that don't work on our CI but on users machines**. However, if a user actually wanted dynamic CRT linkage on a triplet that is for static CRT linkage, they have to configure their triplet accordingly.

If you know a way to not error out by default in this case, I will of curse remove `!staticcrt`!

Please also have a look at the documentation for `vcpkg_check_linkage()`: https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_check_linkage.md

Neumann-A commented:
>  All of the ports I fixed literally have vcpkg_check_linkage(ONLY_DYNAMIC_CRT ONLY_DYNAMIC_LIBRARY)

a) that is a lie. You are also touching ports which have `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` (e.g. hpx)

>  I really don't understand why it should be wrong in this case to express this inside supports

b) `vcpkg_check_linkage` simply existed before `supports`.
c) Your supports statements are syntactical wrong; `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` translates to !static and `vcpkg_check_linkage(ONLY_DYNAMIC_CRT)` translates to !staticcrt; You are doing `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` -> !staticcrt which is wrong and also breaks potential `x64-windows-mt` users. (I don't care about changes to ports using `vcpkg_check_linkage(ONLY_DYNAMIC_CRT ONLY_DYNAMIC_LIBRARY)` but it needs to be checked if it just a limitation of the portfile or incorrect/insufficient build setup.)

> All ports I fixed in this PR produce a build error, which is the reason why supports should be preferred over CI baseline since CI baseline should only be used for things that don't work on our CI but on users machines.

d) supports is meant for what the library supports not for what vcpkg supports through the portfile.
e) ci.baseline.txt is meant to remove build failures `=fail` from ci which are supported by upstream but vcpkg fails to correctly setup. The baseline then becomes basically a TODO list. Or vcpkg needs to introduce a `vcpkg-supports` field which explicitly states what the portfile supports. 

a prime example of a wrong supports statement is the current version of the gstreamer port which should be supported on all platforms but is currently limited to windows for no good reason except whoever made it didn't take the time to fix the portfile for the other platforms......

> If you know a way to not error out by default in this case, I will of curse remove !staticcrt!

`x64-windows-mt`. just explicitly request what the error says.....  (again I don't care about the case  `vcpkg_check_linkage(ONLY_DYNAMIC_CRT [ONLY_DYNAMIC_LIBRARY])`; do what you like with that one.)

Thomas1664 commented:
> x64-windows-mt. just explicitly request what the error says..... (again I don't care about the case vcpkg_check_linkage(ONLY_DYNAMIC_CRT [ONLY_DYNAMIC_LIBRARY]); do what you like with that one.)

Ok. This is something I can work with. I will open another PR for the rest and remove them from this PR

Thomas1664 committed:
revert ONLY_DYNAMIC_LIBRARY cases

Neumann-A commented on ports/libplist/vcpkg.json:
```diff
@@ -1,7 +1,14 @@
   "description": "A library to handle Apple Property List format in binary or XML",
-  "homepage": "https://libimobiledevice.org/"
+  "homepage": "https://libimobiledevice.org/",
+  "supports": "!staticcrt",
```
This is wrong. I went and looked into the blame.
The reason it has `vcpkg_check_linkage(ONLY_DYNAMIC_CRT ONLY_DYNAMIC_LIBRARY)` is because the initial commit of this was using MSBuild with its limitations in regards to setting up different configurations

Neumann-A commented:
I think usage of `vcpkg_check_linkage(ONLY_DYNAMIC_CRT)` needs to be carefully reviewed. For mesa that is ok since it needs to dynamically load other libraries etc but for the rest I doubt it.

Thomas1664 committed:
[treehopper] Fix Linux build

Thomas1664 committed:
[libplist] is supported everywhere

Thomas1664 committed:
version

Thomas1664 committed:
revert baseline

Thomas1664 commented:
> I think usage of `vcpkg_check_linkage(ONLY_DYNAMIC_CRT)` needs to be carefully reviewed. For mesa that is ok since it needs to dynamically load other libraries etc but for the rest I doubt it.

You wrote that you don't care on the `vcpkg_check_linkage(ONLY_DYNAMIC_CRT [ONLY_DYNAMIC_LIBRARY])` case and now you do care?

> (again I don't care about the case `vcpkg_check_linkage(ONLY_DYNAMIC_CRT [ONLY_DYNAMIC_LIBRARY])`;

IMO this is not ok.

Thomas1664 committed:
add private

Thomas1664 committed:
version

Thomas1664 committed:
[treehopper] unsupported on UWP

Thomas1664 committed:
version

Neumann-A commented:
> case and now you do care?

I don't care if `vcpkg_check_linkage(ONLY_DYNAMIC_CRT)` is transformed to `supports: "!staticcrt"` but I do care about **incorrect** usage of  `vcpkg_check_linkage(ONLY_DYNAMIC_CRT)`. That are two different things. 
The first is a valid transformation (which `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` transformed to `supports: "!staticcrt"` was not) while the second one is an invalid added (or missed to remove) `vcpkg_check_linkage(ONLY_DYNAMIC_CRT)`

Thomas1664 committed:
[usbmuxd] Fix build

Thomas1664 committed:
version

Thomas1664 committed:
revert dependencies

Thomas1664 committed:
Fix regressions

Thomas1664 committed:
version

Thomas1664 committed:
Revert moos-core

Thomas1664 committed:
Does treehopper support static crt?

Thomas1664 commented:
@Neumann-A I've checked all ports. All except treehopper are built with MSBuild and treehopper seems to support static CRT as well.
I've split out moos-core in its own PR.

Neumann-A commented:
Hmm then wait until #26370 is merged. I doubt that those CRT checks have any meaning then.

Thomas1664 committed:
revert supports

Thomas1664 committed:
version

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

Thomas1664 committed:
Merge branch 'master' into supports-not-staticcrt
